src/packet/pacer.rs

```diff
@@ -93,53 +94,68 @@ pub trait Pacer {
     ///
     /// **MUST** be called each time [`Pacer::poll_action`] produces [`PollOutcome::PollQueue`] or
     /// [`PollOutcome::PollPadding`]` after the packet is sent.
-    fn register_send(&mut self, now: Instant, packet_size: DataSize, from: QueueId);
+    fn register_send(&mut self, now: Instant, packet_size: DataSize, from: Mid);
 }
 
-/// A unique identifier for a given packet queue. Should be immutable.
-#[derive(Debug, Default, PartialEq, Eq, Hash, Clone, Copy)]
-pub struct QueueId(u64);
-
-/// The state of a single upstream queue.
-/// The pacer manages packets across several upstream queues.
-#[derive(Debug)]
-pub struct QueueState {
-    /// The unique identifier for this queue within the session.
-    pub id: QueueId,
-    /// The total size of queued payloads within the queue.
-    pub size: DataSize,
+#[derive(Debug, Clone, Copy)]
+pub struct QueueSnapshot {
+    /// Time this snapshot was made
+    pub created_at: Instant,
+    /// The total byte size of the snapshot.
+    pub size: usize,
     /// The total number of packets in the queue.
     /// NB: This is not a [`usize`] because it will later be used to divide a [`Duration`], for which
     /// [`usize`] isn't implement. If the queues end up with 2^32 packets something has gone very wrong
     /// in any case.
     pub packet_count: u32,
-    /// The total time the packets in the queue have spent queued.
-    pub total_queue_time: Duration,
-    /// The kind of packets this queue contains.
-    pub queue_kind: MediaKind,
-    /// Whether the media has RTX.
-    pub has_rtx: bool,
-    // The last time a packet was sent from this queue.
-    last_send_time: Option<Instant>,
-    // The time the packet at the front of the queue was queued.
-    leading_packet_queue_time: Option<Instant>,
+    /// Accumulation of all queue time at the time point `created_at`. To use this
+    /// Look at `total_queue_time(now)` which allows getting the queue time at a later Instant.
+    pub total_queue_time_origin: Duration,
+    /// Last time something was emitted from this queue.
+    pub last_emitted: Option<Instant>,
+    /// Time the first unsent packet has spent in the queue.
+    pub first_unsent: Option<Instant>,
+}
+
+impl Default for QueueSnapshot {
+    fn default() -> Self {
+        Self {
+            created_at: not_happening(),
```

Comment:
Wouldn't it be nicer to `QueueSnapshot::new(created_at: Instant)`? No need for this special zero value that way

Reply:
Your logic makes sense with regards to why we couldn't use that constructor, I don't fully agree with regards to the encapsulation bit but w/e